### PR TITLE
Add scoped Mat conversion functions allowing prompt memory freeing.

### PR DESCRIPTION
**Pull request by nh2**
See commit message for details.

In my application where I call out to CUDA while creating `Mat`s,
using `withHMatAsMat` instead of `hMatToMat` changes memory usage
from linearly increasing (depending on whether I force GC) up to 20 GB
to 250 MB constant memory usage.

**Based on #135, please merge that one first.**

### CODE REVIEW

**Comment by nh2** [comment]
This is the cherry-pick / adjustment of my branch [`withHMatAsMat-opencv-0.0.2.1`](https://github.com/nh2/haskell-opencv/tree/withHMatAsMat-opencv-0.0.2.1) for `master`.

File: opencv/src/OpenCV/Internal/C/FinalizerTH.hs [hunk 1]
```diff
@@ -38,6 +38,16 @@ Generated code (stylized):
 > }
 
 > foreign import ccall "&deleteFoo" deleteFoo :: FunPtr (Ptr Foo -> IO ())
+
+And
+
+> mkFinalizer DeletePtr "deleteFoo" "CFoo" ''Foo
```
Comment (basvandijk) [inline opencv/src/OpenCV/Internal/C/FinalizerTH.hs]: `DeletePtr` should be `ReleaseDeletePtr` here.

File: stack.yaml [hunk 2]
```diff
@@ -5,6 +5,9 @@ packages:
 - opencv-extra/
 - opencv-extra-examples/
 flags: {}
+extra-deps:
+# TODO Remove when this is available in the snapshot
+- inline-c-0.7.0.1
```
Comment (basvandijk) [inline stack.yaml]: Could you also add `inline-c >= 0.7.0.0` to the cabal file?

Comment (nh2) [reply stack.yaml]: Done.

File: opencv/src/OpenCV/Internal/C/FinalizerTH.hs (hunk 1 above)
Comment (nh2) [reply opencv/src/OpenCV/Internal/C/FinalizerTH.hs]: Oh right, done.

**Comment by basvandijk** [comment]
@nh2 thanks!

@roelvandijk let's if we can rewrite `mkFinalizer` in terms of [`funPtr`](http://hackage.haskell.org/package/inline-c-0.7.0.1/docs/Language-C-Inline.html#v:funPtr) or maybe we can remove it and use `funPtr` directly inline (like we now do with Mats) which saves a top-level declaration.

**Comment by basvandijk** [comment]
@roelvandijk even better would be to rewrite [`objFromPtr`](https://github.com/LumiGuide/haskell-opencv/blob/master/opencv/src/OpenCV/Internal.hs#L15) in terms of `funPtr`.

**Pull request merged**